joplin-desktop: build from source

fugidev opened the pull request:
Motivation: Upstream doesn't provide binaries for aarch64-linux.

Part of #296939

I took inspiration from the respective packages on [AUR](https://aur.archlinux.org/packages/joplin-desktop) and [Flathub](https://github.com/flathub/net.cozic.joplin_desktop/blob/master/net.cozic.joplin_desktop.yml), which are also source builds.

Unfortunately, building this package is currently broken on Darwin with `sandbox = true` due to #415328 and ~~#416077~~ (merged). Otherwise it does build and work. ~~But because of this, I'd keep the bin package for now and automatically fall back to it on Darwin, to not cause any issues on existing user setups. (Any opinions on this are very welcome.)~~

On Linux, the included "Backup" plugin doesn't load for some reason. The AUR package has the same issue and it has been reported, see https://github.com/JackGruber/joplin-plugin-backup/issues/92. It works on the flatpak tho, and on this package's Darwin build as well.
_Edit: fixed_

More usage testing is greatly appreciated, as well as any other reviews and comments. :)

@yajo you seem to be the only active maintainer of the current package, do you want to be added as maintainer here?

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fugidev committed:
joplin-desktop: build from source

yajo commented on pkgs/by-name/jo/joplin-desktop/package.nix:
```diff
@@ -1,26 +1,187 @@
+  electron,
+  xcbuild,
+  buildPackages,
+  callPackage,
```
I don't think there's any value in having 2 different packages that just return the same app built differently.

If we have a source-code-based packaging option, then we don't need the binary one at all, right?

Even in the case you want to keep it for some time as a fallback in case the new packaging has problems, I'd just add a new argument to the closure like this:

```suggestion
  callPackage,
  packagingMode ? "binary" if stdenv.hostPlatform.isDarwin else "source",
```

Then, inside this same folder, you can have a `package-from-source.nix` and a `package-from-bin.nix`.

fugidev replied:
Thank you for the review. I agree that there is no value in keeping the -bin package once the source built one can fully replace it (i.e. not break Darwin). Unless someone raises another reason to keep it, I'd remove it once the mentioned PR is merged and the other issue fixed.

Maybe putting both packages under `pkgs/by-name/jo/joplin-desktop` like you suggest would be a good idea, yes. I think I'd prefer that to having `joplin-desktop-source`, `joplin-desktop-bin` and an alias `joplin-desktop`, when it's only supposed to be temporary after all.

fugidev replied:
made the change

fugidev replied:
Another thought on this: packages that use `fetchYarnBerryDeps` apparently land on cache.nixos.org for Darwin just fine. So only building it locally, with sandbox enabled, won't work (only if the offlineCache would change even, otherwise e.g. overriding the package should be completely fine too). I think that doesn't actually need to be a blocker for replacing the binary package completely.

The other problem I mentioned is probably still relevant until the fix lands in master: https://nixpk.gs/pr-tracker.html?pr=416077
(Unless hydra builds darwin packages with sandbox disabled)

fugidev commented on pkgs/by-name/jo/joplin-desktop/source/buildPlugin.nix:
```diff
@@ -0,0 +1,50 @@
+      ${lib.getExe npm-lockfile-fix} $out/package-lock.json
+    '';
+  };
+  package = (lib.importJSON (src + "/package.json")) // {
```
I see, thank you.

fugidev commented:
I reworked the plugin building to remove IFDs

yajo commented on pkgs/by-name/jo/joplin-desktop/source/package.nix:
```diff
@@ -0,0 +1,217 @@
+    (makeDesktopItem {
+      name = "joplin";
+      desktopName = "Joplin";
+      exec = "joplin-desktop --no-sandbox %U";
```
question: why `--no-sandbox`?

fugidev replied:
I took all of that from the current package's desktop file, which comes from the AppImage. It seems to be only relevant for the AppImage, see [here](https://github.com/laurent22/joplin/blob/0fec5779321109a0db6af98db9a1237fc28b4b6d/packages/lib/utils/processStartFlags.ts#L134). In the upstream .deb it's not there. I'll remove it.

fugidev commented on pkgs/by-name/jo/joplin-desktop/source/package.nix:
```diff
@@ -0,0 +1,217 @@
+  src = fetchFromGitHub {
+    owner = "laurent22";
+    repo = "joplin";
+    rev = "refs/tags/v${finalAttrs.version}";
```
It is, but it causes an error in nix-prefetch in the update script for some reason: `fetchFromGitHub requires one of either `rev` or `tag` to be provided (not both)`

As a workaround `rev = null` can be added, or extending the nix-prefetch flags with `--rev --expr null`.

I thought the way I did it would be the simplest, but if using `tag` is much preferred then I can of course add one of these workarounds.

fugidev commented:
rebased and updated to 3.3.13

fugidev commented on pkgs/by-name/jo/joplin-desktop/source/buildPlugin.nix:
```diff
@@ -0,0 +1,41 @@
+  patches ? [ ],
+}:
+
+{
```
you're right, it works fine as one function :+1:

fugidev commented:
I refactored buildPlugin.nix as suggested by @yajo.

fugidev commented:
> I'm not sure if this is actually a failure from the new build or from some configuration with local plugins...

Seems like you are running joplin-desktop 2.14.17 there for some reason, not this one

yajo reviewed (approved):
Ah! I tested the wrong PR. Now it worked!

<details><summary>About dialog provided this info FWIW</summary>

Joplin 3.3.13 (prod, linux)

Device: linux, 13th Gen Intel(R) Core(TM) i7-1360P
ID del Cliente: e7cf469d82bd47648320bf15c5040be8
Versión de la Sincronización: 3
Versión del Perfil: 47
Llavero Soportado: No
Alternative instance ID: -

Automatic Backlinks to note: 3.0.3
Freehand Drawing: 3.0.1
Link Graph UI: 1.5.0
Note Tabs: 1.4.0
Outline: 1.5.13
Quick Links: 1.3.2
Rich Markdown: 0.15.0
Toggle Editor Menu & Keyboard Shortcut: 1.0.1

</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2491

r-vdp commented:
I'm not very familiar with yarn packaging, but is there no way to avoid vendoring that huge lock file?
Is there a reason that we can't use the upstream one with the offline cache stuff that I see other packages using?

r-vdp commented:
Hmm, I see now that there is a comment about the vendored lockfile. How would we keep this lockfile up to date? And there is no way to do the needed changes at build time during the patch phase without vendoring the lock file?

fugidev commented:
@r-vdp thank you for taking a look at this.

The yarn.lock is generated and kept up to date by the update script [here](https://github.com/NixOS/nixpkgs/pull/417048/files#diff-8fb3e583580ab1d3ff674a8da8d43f4a5763685ba2a216893b87d38d50957ddfR120).

My reasoning for vendoring it was:
- Using the upstream lockfile *without* deleting unneeded major subpackages causes yarn to install and build additional dependencies that aren't needed for the desktop app. Some of these cause errors (e.g. wasm-pack, required only by the onenote-converter subpackage) that would require additional workarounds to fix. (This causes an increase in size of the offlineCache by ~200MB or 36% btw.)
- Using the upstream lockfile and pruning it in postFetch or in another FOD might be an option (and I would prefer that too), but I unfortunately don't know if `yarn install --mode=update-lockfile` is deterministic. I'm assuming it's probably not.

fugidev commented:
Rebased and updated to v3.4.6 (Note: that's actually a pre-release, but was updated to in #436033).

This required switching to `yarn-berry_4` and some other minor changes.

Build tested on aarch64-linux.

@r-vdp if the size of the vendored yarn.lock is an issue, we could instead store and apply the diff? That would be ~500KB, down from 1.6MB. Or we could try pruning it in the FOD and see if it breaks down the road.

r-vdp commented:
Yeah, the only thing that's keeping me from merging this, is indeed the size of the vendored lock file.

Both of those options sound like they would improve the situation, but I don't know yarn enough to make the call on which approach would be better here.

fugidev commented:
I changed it to store only the diff :+1:
This way, I'm confident that it won't break ^^

fugidev commented:
I got it working! No vendoring a modified lockfile needed anymore. I'm much happier with this.

For some reason I'm getting a hash mismatch for `src` on darwin now, that is very weird...

fugidev commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 417048`
Commit: [`84cecfdf3cff39a731636cb201f2a167a5c223c4`](https://github.com/NixOS/nixpkgs/commit/84cecfdf3cff39a731636cb201f2a167a5c223c4) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/84cecfdf3cff39a731636cb201f2a167a5c223c4..pull/417048/head))
Merge: [`39a7c5b823bdf782fc6edd7dae084eaadc324a9e`](https://github.com/NixOS/nixpkgs/commit/39a7c5b823bdf782fc6edd7dae084eaadc324a9e)

Logs: https://github.com/fugidev/nixpkgs-review-gha/actions/runs/17356630392


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>joplin-desktop</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>joplin-desktop</li>
  </ul>
</details>


---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>joplin-desktop</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>joplin-desktop</li>
  </ul>
</details>

fugidev commented on pkgs/by-name/jo/joplin-desktop/package.nix:
```diff
@@ -1,26 +1,226 @@
+    rev = "refs/tags/v${finalAttrs.version}";
+    postFetch = ''
+      # there's a file with a weird name that causes a hash mismatch on darwin
+      rm $out/packages/app-cli/tests/support/photo*
```
They're not, just test files of a different subpackage we don't use or build anyway 👍

fugidev commented on pkgs/by-name/jo/joplin-desktop/package.nix:
```diff
@@ -1,26 +1,226 @@
+
+    yarn gulp before-dist
+
+    yarn run electronRebuild
+
+    yarn run electron-builder \
+      --dir \
+      -c.electronDist="$electronDist" \
+      -c.electronVersion=${electron.version} \
+      -c.mac.identity=null
```
electron-builder bundles the built files with the specified electron

fugidev commented on pkgs/by-name/jo/joplin-desktop/package.nix:
```diff
@@ -1,26 +1,226 @@
+  src = fetchFromGitHub {
+    owner = "laurent22";
+    repo = "joplin";
+    rev = "refs/tags/v${finalAttrs.version}";
```
See https://github.com/NixOS/nixpkgs/pull/417048#discussion_r2158904303
What would be preferred in your opinion?

r-vdp replied:
Ah, I didn't see that comment, sorry. The `tag` attribute is preferred because it avoids git getting confused when the repo contains a commit whose commit hash has the tag as a prefix.
So I would opt to add `rev = null` with a comment to explain why it is needed for now.

fugidev replied:
`rev = null` didn't work for some reason (it did back then). So I've added the workaround in the update script. Seems more fitting anyway when I think about it, since it's only relevant there.

r-vdp reviewed (approved):
Thanks for all your effort on this!

fugidev commented:
Thank you and everyone else for reviewing and helping get this over the finish line! :)

HHR2020 commented:
This pr makes my joplin scale 4x when NIXOS_OZONE_WL is set.

<img width="2879" height="1710" alt="picture" src="https://github.com/user-attachments/assets/7d8ee8ff-1c9b-47c5-be0e-90a8a2a2611f" />
